hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/lock/OmRequestGatekeeper.java

```diff
@@ -0,0 +1,242 @@
+    for (Lock lock: locks) {
+      if (lock.tryLock(LOCK_TIMEOUT_DEFAULT, TimeUnit.MILLISECONDS)) {
+        try {
+          acquiredLocks.add(lock);
+        } catch (Throwable e) {
+          // We acquired this lock but were unable to add it to our acquired locks list.
+          lock.unlock();
+          releaseLocks(acquiredLocks);
+          throw e;
+        }
+      } else {
+        releaseLocks(acquiredLocks);
+        throw new TimeoutException("Failed to acquire lock after the given timeout.");
+      }
+    }
+  }
+
+  public void unlock(OmLockObject lockObject) {
+    releaseLocks(lockObject.getLocks());
+    lockStatsEnd(lockObject.getLockStats(), lockObject.getLockStatType());
+    lockObject.getLocks().clear();
+  }
+
+  private void releaseLocks(List<Lock> locks) {
+    ListIterator<Lock> reverseIterator = locks.listIterator(locks.size());
+    while (reverseIterator.hasPrevious()) {
+      Lock lock = reverseIterator.previous();
+      lock.unlock();
+    }
+  }
+
+  private static void lockStatsBegin(OmLockStats lockStats, long endTime, long startTime) {
+    lockStats.add(endTime - startTime, OmLockStats.Type.WAIT);
+    lockStats.setLockStartTime(endTime);
+  }
+
+  private static void lockStatsEnd(OmLockStats lockStats, OmLockStats.Type type) {
+    if (lockStats.getLockStartTime() > 0) {
+      lockStats.add(Time.monotonicNowNanos() - lockStats.getLockStartTime(), type);
+    }
+  }
+
+  /**
+   * Lock information after taking locks.
+   */
+  public static class OmLockObject {
+    private final OmLockInfo omLockInfo;
+    private final List<Lock> locks = new ArrayList<>();
+    private final OmLockStats lockStats = new OmLockStats();
+    private OmLockStats.Type lockStatType = OmLockStats.Type.READ;
+
+    public OmLockObject(OmLockInfo lockInfoProvider) {
+      this.omLockInfo = lockInfoProvider;
+    }
+
+    public List<Lock> getLocks() {
+      return locks;
+    }
+
+    public OmLockStats getLockStats() {
+      return lockStats;
+    }
+
+    public OmLockStats.Type getLockStatType() {
+      return lockStatType;
+    }
+
+    public void setLockStatType(OmLockStats.Type lockStatType) {
+      this.lockStatType = lockStatType;
+    }
+
+    public OmLockInfo getOmLockInfo() {
+      return omLockInfo;
+    }
+  }
+
+  /**
+   * lock stats.
+   */
+  public static class OmLockStats {
+    private long lockStartTime;
+    private long waitLockNanos;
+    private long readLockNanos;
+    private long writeLockNanos;
+
+    public long getLockStartTime() {
+      return lockStartTime;
+    }
+
+    public void setLockStartTime(long lockStartTime) {
+      this.lockStartTime = lockStartTime;
+    }
+
+    public long getWaitLockNanos() {
+      return waitLockNanos;
+    }
+
+    public long getReadLockNanos() {
+      return readLockNanos;
+    }
+
+    public long getWriteLockNanos() {
+      return writeLockNanos;
+    }
+
+    void add(long timeNanos, Type type) {
```

Comment:
We can remove the `Type` enum all together and just create one method for updating each stat instead of one method to unpack the enum.

Reply:
done.